src/lib/rule2D.ts

```diff
@@ -0,0 +1,176 @@
+];
+
+/**
+ * Gets the text alternative as defined by one or more native <label>s.
+ * @param elem - element whose text alternative is being calculated
+ * @param context - information relevant to the computation of elem's text alternative
+ * @return - the text alternative for elem if elem is legally labelled by a native
+ * <label>, null otherwise.
+ */
+function getTextIfLabelled(elem: HTMLElement, context: Context): string | null {
+  if (!LABELLABLE_ELEMENT_TYPES.includes(elem.nodeName)) {
+    return null;
+  }
+
+  // Using querySelectorAll to get <label>s in DOM order.
+  const allLabelElems = document.querySelectorAll('label');
+  const labelElems = Array.from(allLabelElems).filter(label => {
+    return (
+      label.htmlFor === elem.id ||
+      // #SPEC_ASSUMPTION (D.2) : Consider the outermost ancestor <label> to
+      // to define the text alternative for elem.
+      (label.contains(elem) && !(label.parentElement?.closest('label') ?? null))
+    );
+  });
+
+  const textAlternative = labelElems
+    .map(labelElem => {
+      return computeTextAlternative(labelElem, {
+        directLabelReference: true,
+        inherited: context.inherited,
+      });
+    })
+    .filter(text => text !== '')
+    .join(' ');
+
+  if (textAlternative !== '') {
+    return textAlternative;
+  }
+
+  return null;
+}
+
+/**
+ * Implementation for rule 2D
+ * @param node - the node whose text alternative is being computed
+ * @param context - information relevant to the text alternative computation
+ * for node
+ * @return - text alternative for node if the conditions for applying
+ * rule 2D are satisfied, null otherwise.
+ */
+export function rule2D(
+  node: Node,
+  context: Context = getDefaultContext()
+): string | null {
+  if (!(node instanceof HTMLElement)) {
+    return null;
+  }
+
+  const roleAttribute = node.getAttribute('role') ?? '';
+  if (roleAttribute === 'presentation' || roleAttribute === 'none') {
+    return null;
+  }
+
+  // #SPEC_ASSUMPTION (D.1) : html-aam (https://www.w3.org/TR/html-aam-1.0/)
+  // specifies all native attributes and elements that define a text alternative.
+
+  const labelText = getTextIfLabelled(node, context);
+  if (labelText) {
+    return labelText;
+  }
+
+  // If input is not <label>led, use native attribute /
+  // element information to compute a text alternative
+  const inputTextAlternative = getUnlabelledInputText(node);
+  if (inputTextAlternative) {
+    return inputTextAlternative;
+  }
+
+  // <caption>s define text alternatives for <table>s
+  if (node instanceof HTMLTableElement) {
+    const captionElem = node.querySelector('caption');
+    if (captionElem) {
+      context.inherited.partOfName = true;
+      return computeTextAlternative(captionElem, {
+        inherited: context.inherited,
+      });
+    }
+  }
+
+  // <figcaption>s define text alternatives for <figure>s
+  // Checking nodeName due to lack of HTMLFigureElement
+  if (node.nodeName.toLowerCase() === 'figure') {
```

Comment:
If `node.nodeName` is in upper case why not leave it as such and compare to `'FIGURE'`?